### DISCUSSION TITLE
I would like to be able to download an export of my org's data instead of having to crawl the API for it.

### DISCUSSION THREAD

**Discussion by StingyJack** [community/community#39535] [Apps, API and Webhooks] [answered]
### Select Topic Area

Product Feedback

### Body

For example if I need to determine average PR time to close and  average number of files per PR across for each team member within each team in our org. There are ~100 teams with between 2 and 6 people on each team (> 400 total developers) and there are about all 2600+ repositories. 

That is a LOT of API requests and after 15-60 minutes the API starts throwing rate limit exceptions. If I could download or export a file with my org's data it would save resources and time for both GH and its users.  

Examples of the data I am referring to would be things like 
 - the list of teams and members
 - the PR's for a repo, including the file names, but not the actual code or code files 
 - repo details like name, creation date, owner, permissions, the file names, but not the actual code or code files

**Comment by LangLangBart** [comment, marked as answer]
> That is a LOT of API requests and after 15-60 minutes the API starts throwing rate limit exceptions.

Did you retrieve the data using only the [REST API](https://docs.github.com/en/rest/guides/getting-started-with-the-rest-api) or did you also try the [GraphQL API](https://docs.github.com/en/graphql/guides/forming-calls-with-graphql)?

---

Your request sound like the perfect use case for a `GraphQL API query`. Something like below, but you need to add `Pagination`. 

- Informative talk on GraphQL by Rea Loretta (15/Dec/19) - [Advanced patterns for GitHub's GraphQL API](https://www.youtube.com/watch?v=i5pIszu9MeM)

> the list of teams and members

:white_check_mark:

> the PR's for a repo, including the file names, but not the actual code or code files

:white_check_mark: (no file names, but only the number of changed files)

> repo details like name, creation date, owner, permissions, the file names, but not the actual code or code files

:white_check_mark: (what do you mean with permission?)

```graphql
{
  rateLimit {
    used
    limit
    cost
  }
  organization(login: "cli") {
    repositories(first: 100, orderBy: {field: STARGAZERS, direction: DESC}) {
      nodes {
        nameWithOwner
        createdAt
        pullRequests(first: 100) {
          totalCount
          nodes {
            changedFiles
            createdAt
            state
            closedAt
            author {
              login
            }
            number
          }
        }
        defaultBranchRef {
          target {
            ... on Commit {
              tree {
                entries {
                  name
                  type
                  object {
                    ... on Tree {
                      entries {
                        name
                        type
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
    teams(first: 100) {
      totalCount
      nodes {
        name
        members(first: 100) {
          totalCount
          nodes {
            login
          }
        }
      }
    }
  }
}
```

```json
{
  "data": {
    "rateLimit": {
      "used": 2,
      "limit": 5000,
      "cost": 2
    },
    "organization": {
      "repositories": {
        "nodes": [
          {
            "nameWithOwner": "cli/cli",
            "createdAt": "2019-10-03T15:24:53Z",
            "pullRequests": {
              "totalCount": 2057,
              "nodes": [
                {
                  "changedFiles": 1,
                  "createdAt": "2019-10-04T03:42:57Z",
                  "state": "MERGED",
                  "closedAt": "2019-10-04T16:01:04Z",
                  "author": {
                    "login": "vilmibm"
                  },
                  "number": 1
                },
                {
                  "changedFiles": 1,
                  "createdAt": "2019-10-07T18:27:37Z",
                  "state": "MERGED",
                  "closedAt": "2019-10-09T20:12:25Z",
                  "author": {
                    "login": "probablycorey"
                  },
                  "number": 3
                },
                ...
```

**Reply by StingyJack** [reply]
Just REST API. I'm looking into the Graph QL interface now but it seems to report incorrect results. Using the example query you have given above but modifying it slightly 
``` graphql
{
  rateLimit {
    used
    limit
    cost
  }
  organization(login: "myOrgLogin") {
    repository(name: "specificRepoName.api") {
      nameWithOwner
      createdAt
      pullRequests(states: CLOSED, orderBy: {field: CREATED_AT, direction: DESC}, first: 100) {
        totalCount
        nodes {
          changedFiles
          createdAt
          state
          closedAt
          author {
            login
          }
          number
        }
      }
    }
  }
}
```
... and it gives me results for the repo that are in the structure above but the `pullRequests.totalCount` is 25 and there are only 25 elements in `pullRequests.nodes`.  There are over 300 pull requests for the repo - well more than 25. 

Any idea why this is unable to return the correct results?  

> what do you mean with permission?

UserA has permission/permissionlevel to push to the repo
TeamB has permission to pull from the repo 
UserB has permission to administer the repo

**Reply by LangLangBart** [reply]
> Any idea why this is unable to return the correct results?

Could you try an array of `CLOSED` and `MERGED`?
<img src="https://user-images.githubusercontent.com/92653266/202963838-420aca0d-6c51-4368-ad95-573a24473a1c.png" width="300">

```graphql
{
  organization(login: "cli") {
    repository(name: "cli") {
      nameWithOwner
      pullRequests(states: [CLOSED, MERGED], orderBy: {field: CREATED_AT, direction: DESC}, first: 100) {
        totalCount
      }
    }
  }
}
```

**Reply by LangLangBart** [reply]
> UserA has permission/permissionlevel to push to the repo
> TeamB has permission to pull from the repo
> UserB has permission to administer the repo

ahhh ... there was recently a discussion about this, but only for the `REST API`.
- https://github.com/community/community/discussions/36528 (18/Oct/22)

#### see the collaborators
I don't know if this will produce the desired results, because I have no test subject. 

<img src="https://user-images.githubusercontent.com/92653266/202968144-32dba8f7-83e6-4e33-907e-9ad13ac28da2.png" width="300">

```graphql
{
  organization(login: "cli") {
    repository(name: "cli") {
      nameWithOwner
      collaborators(affiliation: DIRECT) {
        totalCount
      }
    }
  }
}
```

#### see the members
```graphql
{
  organization(login: "YOUR_ORG") {
    membersWithRole(first: 100) {
      edges {
        node {
          login
        }
        role
      }
    }
  }
}
```
```json
{
  "data": {
    "organization": {
      "membersWithRole": {
        "edges": [
          {
            "node": {
              "login": "USERNAME_01"
            },
            "role": "ADMIN"
          },
          {
            "node": {
              "login": "USERNAME_02"
            },
            "role": "MEMBER"
          },
          {...
```

**Comment by StingyJack** [comment]
The closed/merged thing worked. Thanks. 

I didn't think that GraphQL would represent the same data differently from the REST API, but will keep an eye out for that.
